onnxruntime/core/optimizer/free_dim_override_transformer.cc

```diff
@@ -0,0 +1,88 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/common/logging/logging.h"
+#include "core/session/inference_session.h"
+#include "core/graph/graph_utils.h"
+#include "core/optimizer/free_dim_override_transformer.h"
+
+using namespace ONNX_NAMESPACE;
+using namespace ::onnxruntime::common;
+namespace onnxruntime {
+
+static std::string ToLower(std::string s) {
+  std::transform(s.begin(), s.end(), s.begin(), [](char c) {
+    return static_cast<char>(::tolower(c));
+  });
+
+  return s;
+}
+
+/*explicit*/ FreeDimensionOverrideTransformer::FreeDimensionOverrideTransformer(gsl::span<const FreeDimensionOverride> overrides_to_apply)
+  : GraphTransformer("FreeDimensionOverrideTransformer") {
+  for (const auto& o : overrides_to_apply) {
+    // Convert to lowercase to perform case-insensitive comparisons later
+    std::string denotation = ToLower(o.dimension_denotation);
+
+    dimension_override_by_denotation_.emplace(denotation, o.dimension_override);
+  }
+}
+
+Status FreeDimensionOverrideTransformer::ApplyImpl(Graph& graph, bool& modified, int /*graph_level*/) const {
+  for (const onnxruntime::NodeArg* graph_input : graph.GetInputs()) {
+    // Get the current input's type and shape
+    const auto* input_type = graph_input->TypeAsProto();
+    const auto* input_shape = graph_input->Shape();
+
+    if (!input_type || !input_shape || !input_type->has_tensor_type()) {
+      continue;
+    }
+
+    // Construct a new shape for this input, replacing free dimensions with their overrides
+    onnx::TensorShapeProto new_shape;
+    for (int32_t dim_index = 0; dim_index < input_shape->dim_size(); ++dim_index) {
+      const auto& dimension = input_shape->dim(dim_index);
+
+      // By default just make a copy of the dimension
+      auto* new_dimension = new_shape.add_dim();
+      *new_dimension = dimension;
+
+      if (dimension.has_denotation()) {
+        // Convert to lowercase to perform case-insensitive comparison
```

Comment:
I've typically seen a symbolic dimension specified in dim_param. How does that relate to the usage of denotation? Are they equivalent ways of doing the same thing, in which case we could/should support both?

Reply:
From this [doc](https://github.com/onnx/onnx/blob/master/docs/DimensionDenotation.md#denotation-definition), it doesn't look like this is meant for specifying symbolic shapes. Are the WinML models using this instead of dim_param or both?

Reply:
In WinML, we exclusively use the denotation for overriding free dimensions. As far as I'm aware, the value stored in dim_param isn't really used.

Reply:
If dim_param is used, this optimization won't apply. We should probably check for both?

Reply:
Hmm, I'm not sure. Is that the intended use of dim_param? My understanding is that dim_param can be any arbitrary string, while denotations have defined values (like "DATA_BATCH") that hopefully all converters and tools can rely on. I would lean towards doing the conservative thing right now and using just the denotation. If we find a case where this is causing us to miss an optimization, it's something we can always add later.

Reply:
Based on that document, should something processing denotations stick to looking for the defined names like DATA_BATCH (possibly without the case insensitive matching too for simplicity)?

This also seems like a somewhat rare use case (at least from the models I have seen), so wondering if we really want to add it as a default transformer vs. making it available as part of ORT but having it manually added by people who want to use it. 

The way this transformer works could probably be generalized to support a similar thing with symbolic dimensions where the name to match comes from dim_param, but I'm not sure if there's a real use case for that and I don't think you want a single transformer to handle both. They could share an implementation but handling both in one transformer brings in questions like how to resolve a dimension with a dim_param and denotation.

Reply:
As an example of dim_param usage, the symbolic shape inferencing script used by nuphar to infer shapes just looks at dim_param for symbolic dimension names when propagating shape information through a graph. https://github.com/microsoft/onnxruntime/blob/master/onnxruntime/core/providers/nuphar/scripts/symbolic_shape_infer.py

Reply:
> Based on that document, should something processing denotations stick to looking for the defined names like DATA_BATCH (possibly without the case insensitive matching too for simplicity)?

This is actually how it's implemented in WinML - it looks specifically for the string "DATA_BATCH" and performs the override based on that. But because we've seen models with free dimensions other than the batch, this transformer generalizes the concept to use an arbitrary string. This allows callers to specify any of the defined denotations (e.g. onnx::DATA_FEATURE, onnx::DATA_CHANNEL).

> This also seems like a somewhat rare use case (at least from the models I have seen), so wondering if we really want to add it as a default transformer vs. making it available as part of ORT but having it manually added by people who want to use it. Also 'free dimension override' may be too general a name for something only dealing with denotations.

Interestingly this is something we see quite commonly. @jeffbloo can give more details, but my understanding is that there are several cases where this is important. Most commonly free dimensions used to allow a model to have customizable batch sizes, but I believe we've also seen cases where models are using free width/height dimensions too (e.g. for convolutional models).

The denotation is how free dimensions get tagged, but ultimately what this transformer achieves is that it overrides free dimensions and replaces them with static dimensions. To clarify, this transformer doesn't do anything unless the caller actually specifies any overrides in the session options.

> The way this transformer works could probably be generalized to support a similar thing with symbolic dimensions where the name to match comes from dim_param, but I'm not sure if there's a real use case for that and I don't think you want a single transformer to handle both. They could share an implementation but handling both in one transformer brings in questions like how to resolve a dimension with a dim_param and denotation.

Yes, I think this makes sense. Especially since dim_param and denotation are orthogonal, it seems unusual to look at both with only a single key. Today we have real models that have these denotations and use free dimension overrides in WinML, but I don't think we've seen any equivalent for dim_param.

Reply:
> As an example of dim_param usage, the symbolic shape inferencing script used by nuphar to infer shapes just looks at dim_param for symbolic dimension names when propagating shape information through a graph.

I'm not sure if I understand... the goal of this PR is to replace dim_params with dim_values, so that dim_value is what gets propagated through the graph during shape inference. That is, the goal is to override the free dimensions (dim_param) with static shapes (dim_value) so that the static shapes are available to the operator kernels. This is what allows for the optimizations we can perform at initialization time.